README.md
Description
===========

Install the [Openfire Jabber server](http://www.igniterealtime.org/) from source or rpm.

# Requirements
The following Chef cookbooks should be installed:

* java
* database
* postgresql

## Supported Platforms
* CentOS, Red Hat
* Ubuntu (likely, but untested)

## Database

This *should* still work using the built-in OpenFire database instead of using PostGreSQL. However this needs to be tested.

# Attributes
All attributes are optional

## Version
* `node['openfire']['source_tarball']`: currently defaults to `openfire_4_0_4.tar.gz`
    * This tarball will automatically be downloaded and installed

## Installation
* `node['openfire']['user']`: the local user account to create and use to run the openfire process; defaults to `openfire`
    * also see `node['openfire']['group']`, which also defaults to `openfire`
* `node['openfire']['base_dir']`: the location on the file system to install openfire
* `node['openfire']['config']['admin_console']['port']`: Use your web browser to connect to this port while you are first setting up openfire. Defaults to 9090.
* `node['openfire']['config']['admin_console']['secure_port']`: Use your web browser to connect to this port after you have set up openfire for further configuration. This will require an https/SSL connection. Defaults to 9091.
* `node['openfire']['config']['locale']`: Defaults to `en`.
* `node[:openfire][:config][:network][:interface]`: Defaults to `nil` (listen on all interfaces).

## Database
* `node['openfire']['database']['type']`: currently only works with 'postgresql'. If you want to use the built-in database (untested), do not set this.
* `node['openfire']['database']['password']`: the database password for the Openfire user (required if database type is specified)
* `node['openfire']['database']['name']`: default `openfire`
    * also see `['database']['user']`, `['database']['host']`, `['database']['port']`, which have sane defaults

# Usage

* Optionally set the attributes mentioned in the `Attributes` section.
* Add this to your node's run list: `recipe[openfire]`, then run Chef.
* Startup configuration is in the file `/etc/openfire/openfire.xml`
* Java certificates are in the `/etc/openfire/security` directory.

## New Installation

If you are configuring a new installation of Openfire, use your web browser to connect to your host, port 9091 (or whatever port you chose for `node['openfire']['config']['admin_console']['port']` above). Run through the "wizard", and accept all defaults.

## Import

If you are importing an existing installation of Openfire:

* Optionally import the database
* Within `/etc/openfire/openfire.xml`, right before the line `</jive>`, add the following: `<setup>true</setup>`.

# Upgrading

This cookbook is not yet capable of automatically handling upgrades. To upgrade, follow the [official instructions](http://www.igniterealtime.org/builds/openfire/docs/latest/documentation/upgrade-guide.html).

Download and untar into /opt/openfire. Then set symbolic links:
* `ln -s /etc/openfire /opt/openfire/conf`
* `ln -s /var/log/openfire /opt/openfire/logs`
* `ln -s /etc/openfire/security /opt/openfire/resources/security`

Also copy your plugins:
* rsync -av /opt/openfire_old/plugins/ /opt/openfire/plugins/


**Author: ** Gavin Montague <gavin@leftbrained.co.uk>
**Author: ** Steve Clark <steve@bigsteve.us>  

** Copyright (c) ** 2012-2016 Gavin Montague
** Copyright (c) ** 2016 Steve Clark

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# CHANGELOG for openfire

This file is used to list changes made in each version of openfire.

## 0.2.1
* update url path to a attribute
* add rpm support as the default 

## 0.2.0:

* Add PostGreSQL support via Opscode cookbook
* Configure Openfire XML file as a Chef template including database connection settings
* Add OS init script
* Add Berkshelf support
* Use Java 7 via Opscode cookbook

## 0.1.0:

* Initial release of openfire


test/smoke/default/default_test.rb
 # encoding: utf-8

# Inspec test for recipe tsa_yum::default

# The Inspec reference, with examples and extensive documentation, can be
# found at https://docs.chef.io/inspec_reference.html

describe port(9090) do
  it { should be_listening }
  its('processes') {should include 'java'}
end


recipes/_tarball.rb
include_recipe 'java::default'

if node['openfire']['database']['active']
  include_recipe 'openfire-chef::database'
end

group node['openfire']['group'] do
  system true
end

user node['openfire']['user'] do
  gid node['openfire']['group']
  home node['openfire']['home_dir']
  system true
  shell '/bin/sh'
end

local_tarball_path = "#{Chef::Config['file_cache_path']}/#{node['openfire']['source_tarball']}"

remote_file local_tarball_path do
  checksum node['openfire']['source_checksum']
  source "#{node['openfire']['source_url']}/#{node['openfire']['source_tarball']}"
end

bash 'install_openfire' do
  cwd node['openfire']['base_dir']
  code <<-EOH
    tar xzf #{local_tarball_path}
    chown -R #{node['openfire']['user']}:#{node['openfire']['group']} #{node['openfire']['home_dir']}
    mv #{node['openfire']['home_dir']}/conf /etc/openfire
    rm /etc/openfire/openfire.xml
    mv #{node['openfire']['home_dir']}/logs /var/log/openfire
    mv #{node['openfire']['home_dir']}/resources/security /etc/openfire
  EOH
  creates node['openfire']['home_dir']
end

# link to LSB-recommended directories
link "#{node['openfire']['home_dir']}/conf" do
  to '/etc/openfire'
end

link "#{node['openfire']['home_dir']}/logs" do
  to '/var/log/openfire'
end

link "#{node['openfire']['home_dir']}/resources/security" do
  to '/etc/openfire/security'
end

# this directory contains keys, so lock down its permissions
directory '/etc/openfire/security' do
  group 'openfire'
  mode '0700'
  owner 'openfire'
end

template '/etc/openfire/openfire.xml' do
  action :create_if_missing
  group 'openfire'
  mode '0600'
  owner 'openfire'
end

cookbook_file '/etc/init.d/openfire' do
  mode '0755'
end

# on Debian/Ubuntu we use /etc/default instead of /etc/sysconfig
# make a symlink so that openfirectl is happy
link '/etc/sysconfig' do
  to '/etc/default'
  only_if { node['platform_family'] == 'debian' }
end

template '/etc/sysconfig/openfire' do
  mode '0644'
end

service 'openfire' do
  supports status: true,
           stop: true
  action [:enable, :start]
end

admin_console = node['openfire']['config']['admin_console']
admin_port = admin_console['secure_port'] == -1 ? admin_console['port'] : admin_console['secure_port']
log "And now visit the server on :#{admin_port} to run the openfire wizard." do
  action :nothing
end


attributes/default.rb
default['openfire']['source_tarball'] = 'openfire_4_0_4.tar.gz'
default['openfire']['source_rpm'] = 'openfire-4.1.0-0.2.beta.x86_64.rpm'

default['openfire']['source_url'] = 'http://www.igniterealtime.org/downloadServlet?filename=openfire'


# default to install via rpm
default['openfire']['install_type'] = 'rpm'


default['openfire']['source_checksum'] = '51253420e52c623bee57bfa1494bdfb0aabe7b47e594bb4a4f39e927488d5552'
default['openfire']['rpm_source_checksum'] = '7145cd99ee22dae11ef3c3b1c9f6fc408b33f837c9cfec90d4c6dc5a20423df4'

default['openfire']['base_dir'] = '/opt'
default['openfire']['home_dir'] = "#{openfire['base_dir']}/openfire"
default['openfire']['log_dir'] = '/var/log/openfire'

default['openfire']['user'] = 'openfire'
default['openfire']['group'] = 'openfire'

default['openfire']['pidfile'] = '/var/run/openfire.pid'

# by default, only enable secure admin port
default['openfire']['config']['admin_console']['port'] = 9090
default['openfire']['config']['admin_console']['secure_port'] = 9091

default['openfire']['config']['locale'] = 'en'
default['openfire']['config']['network']['interface'] = nil
